bindings/go/ctf/tar.go

```diff
@@ -0,0 +1,272 @@
+	}()
+
+	// for the extracted version we will first open the CTF with O_RDWR
+	ctf, err := OpenCTFFromOSPath(base, O_RDWR)
+	if err != nil {
+		return nil, fmt.Errorf("unable to setup file system ctf: %w", err)
+	}
+
+	var reader *tar.Reader
+	if format == FormatTGZ {
+		gzipped, err := gzip.NewReader(tarFile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create gzip reader: %w", err)
+		}
+		defer func() {
+			err = errors.Join(err, gzipped.Close())
+		}()
+		reader = tar.NewReader(gzipped)
+	} else {
+		reader = tar.NewReader(tarFile)
+	}
+
+	if err := extractTARToFilesystemCTF(reader, ctf); err != nil {
+		return nil, fmt.Errorf("unable to extract tar to filesystem ctf: %w", err)
+	}
+
+	// if we have an original flag, we will now respect the flag and set the FS to read-only if O_RDONLY is set
+	// this makes sure that even though we just extracted the tar, it can only be read from.
+	if flag&O_RDONLY != 0 || (flag&os.O_WRONLY == 0 && flag&os.O_RDWR == 0) {
+		ctf.FS().ForceReadOnly()
+	}
+
+	return ctf, nil
+}
+
+func extractTARToFilesystemCTF(reader *tar.Reader, ctf *FileSystemCTF) (err error) {
+	for {
+		header, err := reader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if strings.Contains(header.Name, "..") {
+			return fmt.Errorf("invalid tar entry, contains %q: %s", "..", header.Name)
+		}
+		switch header.Typeflag {
+		case tar.TypeReg:
+			if err := ctf.writeFile(header.Name, reader); err != nil {
+				return fmt.Errorf("unable to write file: %w", err)
+			}
+		case tar.TypeDir:
+			if err = os.MkdirAll(header.Name, 0o755); err != nil {
+				return fmt.Errorf("unable to create directory: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Archive creates an archive from the provided CTF and writes it to the specified path.
+// The format of the archive is determined by the format parameter.
+// Supported formats are FormatTAR, FormatTGZ, and FormatDirectory.
+// If the format is FormatDirectory, the filesystem is copied to the specified path.
+func Archive(ctf CTF, path string, format FileFormat) error {
+	switch format {
+	case FormatDirectory:
+		return ArchiveDirectory(ctf, path)
+	case FormatTAR, FormatTGZ:
+		return ArchiveTAR(ctf, path, format)
+	default:
+		return ErrUnsupportedFormat
+	}
+}
+
+// ArchiveDirectory archives the CTF to the specified path with FormatDirectory.
+// The blobs are copied to the directory and the index is written to the index file.
+// The CTF is not modified and only read from.
+// The directory is created if it does not exist.
+// The blobs are written to the blobs directory concurrently.
+func ArchiveDirectory(ctf CTF, path string) error {
+	var fsCTF CTF
+
+	blobs, err := ctf.ListBlobs()
+	if err != nil {
+		return fmt.Errorf("unable to list blobs: %w", err)
+	}
+
+	if fsCTF, err = OpenCTFFromOSPath(path, O_RDWR|O_CREATE); err != nil {
+		return fmt.Errorf("unable to setup file system ctf: %w", err)
+	}
```

Comment:
If I am not mistaken, then `OpenCTFFromOSPath()` creates a new directory for `path` (check the call `filesystem.NewFS(...)`). So, the following `os.Mkdir(...)` should not be necessary, right?

Reply:
fixed, pls double check